chapters/annotations.tex

```diff
@@ -72,6 +72,19 @@ \subsection{Class Description and Revision History}\label{annotation-info-revisi
 <a href="modelica:/MultiBody.Joints.Revolute#info">Revolute</a>
 \end{lstlisting}
 
+\subsubsection{Style Sheet}
+
+A top-level package may contain an annotation \lstinline!styleSheet="modelica:/Package/Resources/style.css"!, which provides a cascading style sheet.
+The given style sheet is used when displaying the \lstinline!info! and \lstinline!revisions! annotations found in the \lstinline!Documentation! annotation.
+The provided style sheet shall not use type or universal selectors.
+\begin{nonnormative}
+It is recommended to use \lstinline[language=CSS]!class! and \lstinline[language=CSS]!id! selectors with a \lstinline[language=CSS]!$NameOfLibrary$-! prefix to avoid collisions when the content is included in a larger context.
+\end{nonnormative}
```

Comment:
```suggestion
\begin{nonnormative}
It is recommended to use \lstinline[language=CSS]!class! and \lstinline[language=CSS]!id! selectors with a \lstinline[language=CSS]!$NameOfLibrary$-! prefix to avoid collisions when the content is included in a larger context.
It is recommended to not use type or universal selectors, unless the tool has documented that possibility.
\end{nonnormative}
```
Since existing style sheets use type and universal selectors (assumedly for good reasons) I don't think we should prohibit it in general. I'm not sure if this should be non-normative or not, but the idea with moving it later was primarily to re-arrange it so that we first have the recommended variant and then the not-so-good one.

Reply:
> I don't think we should prohibit it in general.

I think we should, since a user might see this recommendation, create something that looks good in one version of one tool, but break completely in another tool or tool version.

Reply:
I can understand that is a bit of a problem, but with the existing style sheets used in Dymola we have multiple libraries with e.g.:
```

*       { font-size: ...; font-family: ...; }
...
h5      { font-size: ...; font-weight: ...; color: ...; }
...
h1      { font-size: ...; font-weight: ...; color: ...; }
...
a:link {  color: ...; }
..
```

Reply:
That is exactly the thing that we want to prevent! Overriding that broad of a selection is guaranteed to break styling of tool-generated content.

Reply:
> However, CSS is designed to have precedence between different stylings so it shouldn't just break - but be possible to ensure that you still get the desired result.

Can you explain how this can safeguard tool-generated content? From what I can tell, precedence in CSS is very complicated, and I cannot see how a tool would generate content that is safe from any user-provided CSS.

> And more importantly that is the styling users want to use.

For their own content, surely.

For a page like this, a user should only style the red part:

![image](https://github.com/modelica/ModelicaSpecification/assets/329963/21221c15-d8be-4f8e-bf26-1b79a255ba8f)

The rest, the user cannot be expected to style, since they don't even know what that is, since it may come from a tool they don't even have.

Reply:
> > However, CSS is designed to have precedence between different stylings so it shouldn't just break - but be possible to ensure that you still get the desired result.
> 
> Can you explain how this can safeguard tool-generated content? From what I can tell, precedence in CSS is very complicated, and I cannot see how a tool would generate content that is safe from any user-provided CSS.

My assumption is that the complexity is exactly due to enabling that.

However, most of the precedence can be ignored - users cannot use inline style for the tool-generated elements, and they shouldn't use their IDs; so what matter seems to just be specificity of selectors which simplifies the discussion a lot.

And the simplest solution would be to forbid using `!important` for users, and then use that as an escape mechanism.

Reply:
> And the simplest solution would be to forbid using `!important` for users, and then use that as an escape mechanism.

I guess maybe with some combination of tool usage of `all:revert`? https://developer.mozilla.org/en-US/docs/Web/CSS/all#revert

I don't have enough experience with CSS to tell if this would work.

Reply:
The provided style sheet should not use type or universal selectors, due to possible interference with tool-specific styling.
 (After the non-normative recommendation.)

Reply:
(Language group compromise.)

Reply:
This has now been implemented.  Resolving.